guppylang/checker/expr_checker.py

```diff
@@ -315,10 +316,20 @@ def visit_Name(self, node: ast.Name) -> tuple[ast.Name, Type]:
                 )
             var.used = node
             return with_loc(node, LocalName(id=x)), var.ty
-        elif x in self.ctx.globals.values:
-            # Cache value in the AST
-            value = self.ctx.globals.values[x]
-            return with_loc(node, GlobalName(id=x, value=value)), value.ty
+        elif x in self.ctx.globals:
+            # Look-up what kind of definition it is
+            match self.ctx.globals[x]:
+                case ValueDef() as defn:
+                    return with_loc(node, GlobalName(id=x, def_id=defn.id)), defn.ty
+                # For types, we return their `__new__` constructor
+                case TypeDef() as defn if constr := self.ctx.globals.get_instance_func(
+                    defn, "__new__"
```

Comment:
There is also `int(x)` which takes everthing that implements `__int__`, same for `float(x)` and `bool(x)`.

Before, these were defined as functions, but now the names is already taken by the type.

Python handles this in the same way: `int` is a class/type and "calling" a class executes the constructor

Reply:
Sure, but I'm only worried about the `Qubit` case because adding unnecessary `QAlloc`s (when checking type signatrues for example) will mess up any qubit resource estimates!

However, I'm fairly satisfied we're doing the right thing here

Reply:
Ah I see. Yes, `Name` nodes in types are handled by different code in `tys.parsing`